### PR TITLE
Yielding in get_solver and automatically initialising functions

### LINKED ISSUE #142: Consistent time_partition variable
@jwallwork23, I saw in #116 that you renamed `P = mesh_seq.time_partition` to `tp = mesh_seq.time_partition`. Is this what you'd prefer going forward?

### LINKED ISSUE #179: Initialise fields outside of `get_solver`
Instead of initialising fields inside `get_solver`, I think it would be better to do that somewhere inside `MeshSeq`. That way we do not require the user to properly name the functions (`"field"` and `"field_old"`) and assigning initial conditions to `field_old`. I think it would be good to hide this inner working of goalie from the user.

So, for example, the `burgers.py` demo is currently:
```
def get_solver(mesh_seq):
    def solver(index, ic):
        function_space = mesh_seq.function_spaces["u"][index]
        u = Function(function_space, name="u")

        # Initialise 'lagged' solution
        u_ = Function(function_space, name="u_old")
        u_.assign(ic["u"])
        ...
```
but would be simplified to
```
def get_solver(mesh_seq):
    def solver(index):                  # no longer passing ic
        u, u_ = mesh_seq.fields["u"]    # now mesh_seq.fields returns a dictionary
        ...
```

What do you think @jwallwork23?

### PR DESCRIPTION
**Pull request by ddundo**
This PR partially addresses #145: `get_solver` functions should now `yield` at every timestep instead of returning the solution dictionary at the end of the time integration. This allows us to avoid annotating to tape in `solve_forward` (we still can if we want to in the future for some reason). So everything tape-related has been moved to `adjoint.py`. Yielding also gives us a lot of flexibility, so I introduced a private method `MeshSeq._solve_forward` which is now called from `MeshSeq.solve_forward` and `MeshSeq.get_checkpoints`, which previously had separate implementations.

### CODE REVIEW

File: goalie/mesh_seq.py
Comment (ddundo) [inline goalie/mesh_seq.py:365]: I forgot to call `next(solver_gen)` here

File: goalie/mesh_seq.py
Comment (ddundo) [inline goalie/mesh_seq.py:495]: Fix docstring

File: goalie/mesh_seq.py
Comment (ddundo) [inline goalie/mesh_seq.py:478]: Fix docstring

File: demos/point_discharge2d.py
Comment (jwallwork23) [inline demos/point_discharge2d.py:149]: I notice that in some cases index 0 is used and in others index 1. Is it the case that for steady-state problems index 0 is used and index 1 for time-dependent?

File: test_adjoint/test_adjoint.py
Comment (jwallwork23) [inline test_adjoint/test_adjoint.py:153]: I assume this isn't meant to be here?

File: goalie/mesh_seq.py
Comment (jwallwork23) [inline goalie/mesh_seq.py:51]: Is the intention for this to hold non-prognostic fields as well? I'm conscious that with this change we will be using the term `fields` in place of `solutions`.

Also, if we do adopt `MeshSeq.fields` then `TimePartition.fields` should probably be renamed for clarity, e.g., `TimePartition.field_names`.

File: goalie/mesh_seq.py
Comment (jwallwork23) [inline goalie/mesh_seq.py:365]: I think these kind of queries should only be done in debugging mode.

Also, minor point: capitalise first character of warning message for consistency with other warnings and errors.

**Review by jwallwork23 — COMMENTED** [review]
Some initial comments ahead of discussing in today's meeting.

One general note is that I notice you've thrown in quite a few unrelated formatting changes. Its okay to throw in one or two if they involve the code you are already changing, but please refrain from doing this extensively in future, especially for unrelated changes and if the PR is already large.

File: goalie/mesh_seq.py
Comment (jwallwork23) [inline goalie/mesh_seq.py:485]: This is really nice!

Formatting: you can use `for field, initial_condition in enumerate(initial_conditions):`.

File: demos/point_discharge2d.py
Comment (ddundo) [reply demos/point_discharge2d.py:149]: I will get rid of this confusion in a very near-future PR, so that we do not even initialise the lagged solution for steady problems

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py:51]: That's a good point. The dictionary `mesh_seq.fields` does exactly what you said: holds the solutions at the current timestep. Should we rename it to something more descriptive of that?

Comment (ddundo) [reply goalie/mesh_seq.py:51]: But it should also be distinct from the `solutions` in `function_data.py`

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py:365]: I realise now that `"solution unchanged after first timestep."` is also a bad description... Will fix

File: test_adjoint/test_adjoint.py
Comment (ddundo) [reply test_adjoint/test_adjoint.py]: Sorry about that! Removed in a29c3e38d5b307cb5c762f84071c16ace78c5204

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py]: Actually I agree - sorry for complicating. Done as you suggested in 2737ad669b93f9a340b2799aa5351e08ceb36a65

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py:51]: With my analogy between `MeshSeq.function_spaces` and `MeshSeq.fields` that I described in https://github.com/mesh-adaptation/goalie/issues/179#issuecomment-2079326965, maybe we could rename `MeshSeq.fields` to `MeshSeq.functions`? I think that fits nicely

File: goalie/mesh_seq.py
Comment (ddundo) [inline goalie/mesh_seq.py:507]: Actually solutions should be reinitialised only if `update_solutions=True` I guess

**Comment by ddundo** [comment]
After separating out #179, this is now again ready for review. Sorry for the messy commit history - it was a pain to rebase, so I did it semi-manually with `git checkout main --`.

I think this should be easy to review now. Most of the changes are due to copy-pasting tape-related methods from `mesh_seq.py` to `adjoint.py`.

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py]: Done in c86222b340fcdd039c178b1b1adc79544fc0deb1

File: goalie/mesh_seq.py
Comment (jwallwork23) [inline goalie/mesh_seq.py:509]: Not sure if it's standard practice but you could put
```
:yields: ...
:ytype: ...
```

File: goalie/mesh_seq.py
Comment (jwallwork23) [inline goalie/mesh_seq.py:378]: Any new TODOs need to reference issues so you'll need to create one and mention it here before this gets merged.

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
This looks good, but did we actually end up agreeing that yielding in this way was a good idea? There were concerns that it would make it difficult to couple Goalie to other codes whose solvers don't use such an approach.

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py:378]: Oh sorry, I actually meant to address this in this PR! I forgot

**Comment by ddundo** [comment]
> This looks good, but did we actually end up agreeing that yielding in this way was a good idea? There were concerns that it would make it difficult to couple Goalie to other codes whose solvers don't use such an approach.

I thought @stephankramer's concerns about coupling were related to #179 and not this. But yes, probably best to discuss together :) and the main point of this PR is to avoid memory issues (i.e. avoid taping) so maybe there are better options altogether

**Review by stephankramer — APPROVED** [review]
This is really nice. Yes my objections where more to do with the `mesh_seq.fields` changes which were actually more generally to do with the opaqueness of the data flow between goalie and user code, which I'm sure we'll address at some point :-)

This looks like a great improvement alround. There's TODO's left, that @jwallwork23 pointed at, otherwise afaic this is good to go!

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py]: Thanks! Done in c0abbab024ffc394978cef64dfc56b464adb2132

File: goalie/mesh_seq.py
Comment (ddundo) [reply goalie/mesh_seq.py]: Fixed in c0abbab024ffc394978cef64dfc56b464adb2132

**Comment by ddundo** [comment]
Thanks both for reviewing this! I made the requested changes :)

@acse-ej321 mentioned today that this didn't work when using Runge-Kutta, since each computed stage should be written on tape. The changes here shouldn't hinder taping - I just paused it inside `MeshSeq._solve_forward`. So I assume you need to do `pyadjoint.continue_annotation()` in your script?

**Review by jwallwork23 — APPROVED** [review]

**Comment by jwallwork23** [comment]
> @acse-ej321 mentioned today that this didn't work when using Runge-Kutta, since each computed stage should be written on tape.

Goalie itself doesn't work for general RK methods. The way it's currently designed, it will only work for theta methods.

**Comment by ddundo** [comment]
> Goalie itself doesn't work for general RK methods. The way it's currently designed, it will only work for theta methods.

I may have misunderstood then - I thought it was working on the main branch

**Comment by jwallwork23** [comment]
Oh okay. I may have misremembered. Will look into it.

**Comment by jwallwork23** [comment]
> I may have misunderstood then - I thought it was working on the main branch

@ddundo which RK methods did you say worked for you?

Okay I've refreshed my memory. Goalie basically takes Pyadjoint apart and stiches it back together with a few assumptions. We assume solutions can be found in the `adj_sol` attribute of a `SolveBlock` and that the solution at the previous timestep is a dependency of the `SolveBlock`. However, what if you have a multi-stage method with tendencies, etc.? In that case surely the `SolveBlock` will depend on a tendency rather than a lagged solution field. We don't currently account for this. I'd be happy to be proven wrong, of course.

**Comment by jwallwork23** [comment]
I realise this discussion is tangential to the PR here. Is it good to merge? If so, maybe open an issue for RK and continue the discussion there?

**Comment by ddundo** [comment]
Thanks for looking into this! But sorry if I wasn't clear - it was @acse-ej321 who said she was using RK - so she'll be able to answer this :)

**Comment by ddundo** [comment]
> I realise this discussion is tangential to the PR here. Is it good to merge? If so, maybe open an issue for RK and continue the discussion there?

Well if I understood correctly, @acse-ej321 couldn't get RK to work with yielding. So I think I should wait to hear from her before merging

**Comment by acse-ej321** [comment]
@jwallwork23 Apologies for the confusion. I likely mixed two topics up in the last meeting. I had asked @ddundo about multi-stage solvers in the previous meeting as a separate topic. I also mentioned that I could not run my current workflow between Goalie and Thetis with the #182 branch. The current workflow is built on Thetis writing to tape for the forward solve. In a recent conversation, @stephankramer mentioned there might need to be some additional functionality added to Thetis for `yield` to integrate well with this PR.

**Comment by ddundo** [comment]
Thanks for clarifying @acse-ej321!

I'm still a bit confused if this should then be addressed on the Goalie side (in this PR) :) but it sounds like a simple solution is to add a kwarg to `MeshSeq.solve_forward` to prevent calling `pyadjoint.pause_annotation()`?

**Comment by ddundo** [comment]
I'll go ahead and merge this. Please open an issue and feel free to assign me if needed :)

**Pull request merged**
